#5829 broke parallel builds with gnu make.

As of #5829 parallel builds are randomly broken, because add_custom_command doesn't parallel well.
```
[  0%] Generating include/SDL.h
[  0%] Generating include/close_code.h
[  0%] Generating include/SDL_assert.h
[  0%] Generating include/close_code.h
[  0%] Generating include/SDL.h
[  0%] Generating include/SDL.h
[  0%] Generating include/SDL.h
[  0%] Generating include/SDL_atomic.h
[  0%] Generating include/SDL_atomic.h
[  0%] Generating include/SDL_assert.h
[  0%] Generating include/SDL_audio.h
[  0%] Generating include/SDL_assert.h
[  0%] Generating include/SDL_audio.h
[  0%] Generating include/SDL_atomic.h
[  1%] Generating include/SDL_bits.h
[  1%] Generating include/SDL_atomic.h
[  1%] Generating include/SDL_audio.h
[  1%] Generating include/SDL_blendmode.h
[  1%] Generating include/SDL_bits.h
[  1%] Generating include/SDL_clipboard.h
[  1%] Generating include/SDL_bits.h
[  3%] Generating include/SDL_copying.h
[  3%] Generating include/SDL_audio.h
[  3%] Generating include/SDL_blendmode.h
[  3%] Generating include/SDL_cpuinfo.h
[  3%] Generating include/SDL_blendmode.h
[  3%] Generating include/SDL_copying.h
[  3%] Generating include/SDL_blendmode.h
[  3%] Generating include/SDL_clipboard.h
[  3%] Generating include/SDL_cpuinfo.h
[  3%] Generating include/SDL_egl.h
[  3%] Generating include/SDL_copying.h
[  3%] Generating include/SDL_cpuinfo.h
[  3%] Generating include/SDL_egl.h
[  3%] Generating include/SDL_egl.h
[  3%] Generating include/SDL_endian.h
[  3%] Generating include/SDL_endian.h
[  3%] Generating include/SDL_egl.h
[  3%] Generating include/SDL_endian.h
[  3%] Generating include/SDL_error.h
[  5%] Generating include/SDL_error.h
[  5%] Generating include/SDL_error.h
[  5%] Generating include/SDL_events.h
[  5%] Generating include/SDL_events.h
[  5%] Generating include/SDL_events.h
[  5%] Generating include/SDL_endian.h
[  5%] Generating include/SDL_events.h
[  5%] Generating include/SDL_filesystem.h
[  5%] Generating include/SDL_gamecontroller.h
[  5%] Generating include/SDL_filesystem.h
[  5%] Generating include/SDL_filesystem.h
[  5%] Generating include/SDL_gamecontroller.h
[  5%] Generating include/SDL_gamecontroller.h
[  5%] Generating include/SDL_filesystem.h
[  7%] Generating include/SDL_gamecontroller.h
[  7%] Generating include/SDL_gesture.h
[  7%] Generating include/SDL_guid.h
[  7%] Generating include/SDL_gesture.h
[  7%] Generating include/SDL_gesture.h
[  7%] Generating include/SDL_gesture.h
[  7%] Generating include/SDL_haptic.h
[  7%] Generating include/SDL_guid.h
[  7%] Generating include/SDL_guid.h
[  7%] Generating include/SDL_haptic.h
[  7%] Generating include/SDL_haptic.h
[  7%] Generating include/SDL_guid.h
[  7%] Generating include/SDL_hidapi.h
[  7%] Generating include/SDL_hidapi.h
[  7%] Generating include/SDL_haptic.h
[  7%] Generating include/SDL_hidapi.h
[  7%] Generating include/SDL_hidapi.h
[  8%] Generating include/SDL_hints.h
[  8%] Generating include/SDL_joystick.h
[  8%] Generating include/SDL_joystick.h
[  8%] Generating include/SDL_joystick.h
[  8%] Generating include/SDL_hints.h
[  9%] Generating include/SDL_hints.h
[  9%] Generating include/SDL_hints.h
[  9%] Generating include/SDL_keycode.h
[  9%] Generating include/SDL_keyboard.h
[  9%] Generating include/SDL_joystick.h
[ 10%] Generating include/SDL_keyboard.h
[ 10%] Generating include/SDL_keyboard.h
[ 10%] Generating include/SDL_keycode.h
[ 10%] Generating include/SDL_loadso.h
[ 10%] Generating include/SDL_loadso.h
[ 10%] Generating include/SDL_keyboard.h
[ 10%] Generating include/SDL_locale.h
[ 10%] Generating include/SDL_loadso.h
[ 10%] Generating include/SDL_keycode.h
[ 10%] Generating include/SDL_locale.h
[ 10%] Generating include/SDL_locale.h
[ 11%] Generating include/SDL_loadso.h
[ 11%] Generating include/SDL_log.h
[ 11%] Generating include/SDL_log.h
[ 11%] Generating include/SDL_main.h
[ 11%] Generating include/SDL_main.h
[ 11%] Generating include/SDL_log.h
[ 11%] Generating include/SDL_locale.h
[ 11%] Generating include/SDL_messagebox.h
[ 11%] Generating include/SDL_log.h
[ 11%] Generating include/SDL_main.h
[ 12%] Generating include/SDL_main.h
[ 12%] Generating include/SDL_messagebox.h
[ 12%] Generating include/SDL_metal.h
[ 13%] Generating include/SDL_metal.h
[ 13%] Generating include/SDL_messagebox.h
[ 13%] Generating include/SDL_messagebox.h
[ 13%] Generating include/SDL_misc.h
[ 13%] Generating include/SDL_metal.h
[ 14%] Generating include/SDL_misc.h
[ 14%] Generating include/SDL_metal.h
[ 14%] Generating include/SDL_mouse.h
[ 14%] Generating include/SDL_mouse.h
[ 14%] Generating include/SDL_misc.h
[ 14%] Generating include/SDL_mutex.h
[ 14%] Generating include/SDL_name.h
[ 14%] Generating include/SDL_mutex.h
[ 14%] Generating include/SDL_mouse.h
[ 14%] Generating include/SDL_misc.h
[ 14%] Generating include/SDL_opengl.h
[ 14%] Generating include/SDL_name.h
[ 14%] Generating include/SDL_opengl_glext.h
[ 14%] Generating include/SDL_mutex.h
[ 15%] Generating include/SDL_mutex.h
[ 15%] Generating include/SDL_opengl.h
[ 16%] Generating include/SDL_opengl.h
[ 16%] Generating include/SDL_opengl_glext.h
[ 16%] Generating include/SDL_opengles.h
[ 16%] Generating include/SDL_opengl_glext.h
[ 16%] Generating include/SDL_opengles.h
[ 16%] Generating include/SDL_opengl.h
[ 16%] Generating include/SDL_opengles2.h
[ 16%] Generating include/SDL_opengles.h
[ 16%] Generating include/SDL_opengles2.h
[ 16%] Generating include/SDL_opengles2.h
[ 16%] Generating include/SDL_opengles2_gl2.h
[ 17%] Generating include/SDL_opengles.h
Error copying file (if different) from "/home/isage/dev/vita/mine/SDL/include/SDL_opengles2.h" to "/home/isage/dev/vita/mine/SDL/build_pc/include/SDL_opengles2.h".
make[2]: *** [CMakeFiles/SDL2-static.dir/build.make:266: include/SDL_opengles2.h] Error 1
make[2]: *** Deleting file 'include/SDL_opengles2.h'
make[1]: *** [CMakeFiles/Makefile2:141: CMakeFiles/SDL2-static.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 17%] Generating include/SDL_opengles2.h
[ 17%] Generating include/SDL_opengles2_gl2ext.h
[ 17%] Generating include/SDL_opengles2_gl2ext.h
[ 17%] Generating include/SDL_opengles2_gl2.h
[ 17%] Generating include/SDL_opengles2_gl2platform.h
[ 18%] Generating include/SDL_opengles2_gl2ext.h
^Cmake[2]: *** Deleting file 'include/SDL_opengles2_gl2.h'
```
as you can see, it can run multiple times, creating race-condition, when two (or more) build jobs try to run same command simultaneously.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -109,6 +109,10 @@ math(EXPR DYLIB_COMPAT_VERSION_MAJOR "${LT_MAJOR} + 1")
 set(DYLIB_CURRENT_VERSION "${DYLIB_CURRENT_VERSION_MAJOR}.${DYLIB_CURRENT_VERSION_MINOR}.0")
 set(DYLIB_COMPATIBILITY_VERSION "${DYLIB_COMPAT_VERSION_MAJOR}.0.0")
 
+# This list holds all generated headers.
+# To avoid generating them twice, these are added to a dummy target on which all sdl targets depend.
+set(SDL_GENERATED_HEADERS)
+
 #message(STATUS "${LT_VERSION} :: ${LT_AGE} :: ${LT_REVISION} :: ${LT_CURRENT} :: ${LT_RELEASE}")
 
 # General settings & flags
@@ -2716,7 +2720,10 @@ endif()
 # 1. replace all `#cmakedefine`'s and `@abc@`
 configure_file("${SDL2_SOURCE_DIR}/include/SDL_config.h.cmake"
   "${SDL2_BINARY_DIR}/SDL_config.h.intermediate")
-# 2. generate SDL_config in an build_type-dependent folder (which should be first in the include search path)
+# 2. Create the "include-config-${CMAKE_BUILD_TYPE}" folder (fails on older CMake versions when it does not exist)
+string(TOLOWER "${CMAKE_BUILD_TYPE}" lower_build_type)
+execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}/include-config-${lower_build_type}")
+# 3. generate SDL_config in an build_type-dependent folder (which should be first in the include search path)
 file(GENERATE
     OUTPUT "${SDL2_BINARY_DIR}/include-config-$<LOWER_CASE:$<CONFIG>>/SDL_config.h"
     INPUT "${SDL2_BINARY_DIR}/SDL_config.h.intermediate")
@@ -2779,6 +2786,7 @@ foreach(_hdr IN LISTS SDL2_INCLUDE_FILES)
         DEPENDS "${_hdr}")
   endif()
 endforeach()
+list(APPEND SDL_GENERATED_HEADERS ${SDL2_COPIED_INCLUDE_FILES})
 
 if(NOT WINDOWS OR CYGWIN OR MINGW)
 
@@ -2896,6 +2904,10 @@ if (SDL_ASAN)
   endif()
 endif()
 
+# Create target that collects all all generated include files.
+add_custom_target(sdl_headers_copy
+    DEPENDS ${SDL_GENERATED_HEADERS})
+
 ##### Info output #####
 message(STATUS "")
 message(STATUS "SDL2 was configured with the following options:")
@@ -2967,7 +2979,8 @@ set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${EXTRA_CFLAGS} ${EXTRA_CFLAGS_BUILD}")
 
 if(NOT WINDOWS_STORE AND NOT SDL2_DISABLE_SDL2MAIN)
   # Build SDLmain
-  add_library(SDL2main STATIC ${SDLMAIN_SOURCES} ${SDL2_COPIED_INCLUDE_FILES})
+  add_library(SDL2main STATIC ${SDLMAIN_SOURCES})
+  add_dependencies(SDL2main sdl_headers_copy)
   # alias target for in-tree builds
   add_library(SDL2::SDL2main ALIAS SDL2main)
   target_include_directories(SDL2main BEFORE PRIVATE "${SDL2_BINARY_DIR}/include" PRIVATE "${SDL2_BINARY_DIR}/include-config-$<LOWER_CASE:$<CONFIG>>")
@@ -2997,7 +3010,8 @@ if(APPLE)
 endif()
 
 if(SDL_SHARED)
-  add_library(SDL2 SHARED ${SOURCE_FILES} ${VERSION_SOURCES} ${SDL2_COPIED_INCLUDE_FILES})
+  add_library(SDL2 SHARED ${SOURCE_FILES} ${VERSION_SOURCES})
+  add_dependencies(SDL2 sdl_headers_copy)
   # alias target for in-tree builds
   add_library(SDL2::SDL2 ALIAS SDL2)
   set_target_properties(SDL2 PROPERTIES POSITION_INDEPENDENT_CODE TRUE)
@@ -3055,7 +3069,8 @@ if(SDL_SHARED)
 endif()
 
 if(SDL_STATIC)
-  add_library(SDL2-static STATIC ${SOURCE_FILES} ${SDL2_COPIED_INCLUDE_FILES})
+  add_library(SDL2-static STATIC ${SOURCE_FILES})
+  add_dependencies(SDL2-static sdl_headers_copy)
   # alias target for in-tree builds
   add_library(SDL2::SDL2-static ALIAS SDL2-static)
   if(MSVC OR (WATCOM AND (WIN32 OR OS2)))
@@ -3097,7 +3112,8 @@ set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DSDL_BUILD_MICRO_VERSION=${SDL_MICRO_VERSIO
 
 if(SDL_TEST)
   file(GLOB TEST_SOURCES ${SDL2_SOURCE_DIR}/src/test/*.c)
-  add_library(SDL2_test STATIC ${TEST_SOURCES} ${SDL2_COPIED_INCLUDE_FILES})
+  add_library(SDL2_test STATIC ${TEST_SOURCES})
+  add_dependencies(SDL2_test sdl_headers_copy)
   add_library(SDL2::SDL2test ALIAS SDL2_test)
   set_target_properties(SDL2_test PROPERTIES
       EXPORT_NAME SDL2test)

--- a/cmake/sdlchecks.cmake
+++ b/cmake/sdlchecks.cmake
@@ -632,6 +632,7 @@ macro(WaylandProtocolGen _SCANNER _CODE_MODE _XML _PROTL)
         ARGS "${_CODE_MODE}" "${_XML}" "${_WAYLAND_PROT_C_CODE}"
     )
 
+    list(APPEND SDL_GENERATED_HEADERS "${_WAYLAND_PROT_H_CODE}")
     list(APPEND SOURCE_FILES "${_WAYLAND_PROT_C_CODE}")
 endmacro()
 

